test/extended/util/metrics.go

```diff
@@ -0,0 +1,352 @@
+	podInfos := []*podInfo{}
+	for _, pod := range pods.Items {
+		info := &podInfo{
+			name:      pod.Name,
+			namespace: pod.Namespace,
+			status:    string(pod.Status.Phase),
+		}
+
+		podReady, err := utils.PodRunningReady(&pod)
+		if !podReady || err != nil {
+			result := "skipped, pod is not Running"
+			if err != nil {
+				result = fmt.Sprintf("%s: %v", result, err)
+			}
+
+			info.result = result
+			info.skipped = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		if len(pod.Spec.Containers) == 0 {
+			info.result = "skipped, no containers found"
+			info.skipped = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		metrics, err := getPodMetrics(adminConfig, &pod, bearerToken)
+		if err != nil {
+			// ignore errors from pods with no /metrics endpoint available
+			if !strings.Contains(err.Error(), "Connection refused") {
+				info.result = fmt.Sprintf("error: %s", err)
+				info.failed = true
+			} else {
+				info.skipped = true
+				info.result = fmt.Sprintf("/metrics endpoint not available")
+			}
+
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		metricGroup, ok := metrics["rest_client_requests_total"]
+		if !ok {
+			info.result = fmt.Sprintf("error: failed to find counter: %q", "rest_client_requests_total")
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		procStartTime, ok := metrics["process_start_time_seconds"]
+		if !ok || len(procStartTime.Metric) == 0 {
+			info.result = fmt.Sprintf("error: failed to find metric: %q", "process_start_time_seconds")
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+		procStartTimeSeconds := procStartTime.Metric[0].GetGauge().GetValue()
+		totalProcTimeSeconds := time.Now().Unix() - int64(procStartTimeSeconds)
+
+		totalRequestCount := float64(0)
+		for _, metric := range metricGroup.Metric {
+			totalRequestCount += metric.Counter.GetValue()
+		}
+
+		comment := "within QPS bounds"
+		qps := totalRequestCount / float64(totalProcTimeSeconds)
+		if qps > maxQPSAllowed {
+			comment = "exceeds QPS bounds"
+		}
+		info.status = fmt.Sprintf("%s (%s)", info.status, comment)
+		info.qps = qps
+		info.result = fmt.Sprintf("%v requests over a span of %v seconds", totalRequestCount, totalProcTimeSeconds)
+		podInfos = append(podInfos, info)
+	}
+
+	return podInfos, nil
+}
+
+func getBearerToken(adminClient kubernetes.Interface, namespace string) (string, error) {
+	if len(calculatedBearerToken) > 0 {
+		return calculatedBearerToken, nil
+	}
+
+	sa := &v1.ServiceAccount{
+		TypeMeta:   metav1.TypeMeta{APIVersion: v1.SchemeGroupVersion.String(), Kind: "ServiceAccount"},
+		ObjectMeta: metav1.ObjectMeta{Name: namespace + "-sa", Namespace: namespace},
+	}
+
+	// ensure our namespace exists
+	_, err := adminClient.CoreV1().Namespaces().Create(&v1.Namespace{
+		TypeMeta:   metav1.TypeMeta{APIVersion: v1.SchemeGroupVersion.String(), Kind: "Namespace"},
+		ObjectMeta: metav1.ObjectMeta{Name: namespace},
```

Comment:
test namespaces must have generated name or you will fail on shared clusters or if you keep namespaces from previous runs. also verify if regular test namespaces have some label (I think we have a bug and we drop the labels at the project request so even though we set those in code you won't see them on the namespace created.)

Reply:
there is generateName in metav1